manager/network/cni/cni.go

```diff
@@ -0,0 +1,84 @@
+package cni
+
+import (
+	"github.com/containernetworking/cni/libcni"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/allocator/cniallocator"
+	"github.com/docker/swarmkit/manager/network"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+type cni struct{}
+
+// New produces a fresh CNI Network Model
+func New() network.Model {
+	return &cni{}
+}
+
+func (nm *cni) NewAllocator() (network.Allocator, error) {
+	return cniallocator.New()
+}
+
+func (nm *cni) SupportsIngressNetwork() bool {
+	return false
+}
+
+func parseCNIspec(spec *api.NetworkSpec) (*libcni.NetworkConfig, error) {
+	// This is rather similar to cniConfig in the containerd executor...
+	cniConfig, ok := spec.DriverConfig.Options["config"]
+	if !ok {
+		return nil, grpc.Errorf(codes.InvalidArgument, "CNI network has no config")
+	}
+
+	cni, err := libcni.ConfFromBytes([]byte(cniConfig))
+	if err != nil {
+		return nil, grpc.Errorf(codes.InvalidArgument, "Failed to parse CNI config: %s", err)
+	}
+	return cni, nil
+}
+
+func (nm *cni) SetDefaults(spec *api.NetworkSpec) error {
+	if spec.DriverConfig == nil || spec.DriverConfig.Name != "cni" {
+		return nil
+	}
+
+	cni, err := parseCNIspec(spec)
+	if err != nil {
+		return err
+	}
+
+	if spec.Annotations.Name == "" {
+		spec.Annotations.Name = cni.Network.Name
```

Comment:
> I'm not sure I understand why spec.Annotations.Name would have to match the CNI name

Just to avoid confusion since the CNI name might get used where the admin might see it e.g. the bridge name (I think) and having that match swarmd's idea of the network's name seems useful.

Maybe it's not worth enforcing it though, if an admin wants to confuse themself that's up to them.